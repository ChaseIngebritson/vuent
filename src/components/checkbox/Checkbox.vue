<template>
  <label class="vnt-checkbox">
    <input class="vnt-checkbox__input"
           type="checkbox"
           :checked="checked"
           :disabled="disabled" />
    <span class="vnt-checkbox__tick"></span>
    <span class="vnt-checkbox__text">
      <slot>{{ label }}</slot>
    </span>
  </label>
</template>

<script>
export default {
  name: 'VntCheckbox',

  props: {
    label: {
      type: String,
      default: 'Checkbox'
    },
    checked: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    }
  }
};
</script>

<style lang="scss">
@import '../../scss/mixins/component';

.vnt-checkbox {
  @include component-base();

  position: relative;
  display: flex;
  align-items: center;
  line-height: 30px;
}

.vnt-checkbox__input {
  position: absolute;
  clip-path: inset(20px 20px);

  &:checked + .vnt-checkbox__tick {
    background: $accentColor;
    border-color: $accentColor;
  }

  &:disabled + .vnt-checkbox__tick {
    border-color: #999;
  }

  &:disabled:checked + .vnt-checkbox__tick {
    background: transparent;

    &:after {
      border-color: #999;
    }
  }
}

.vnt-checkbox__tick {
  background: transparent;
  border: 2px solid #000;
  box-sizing: border-box;
  width: 20px;
  height: 20px;
  position: relative;

  &:after {
    content: '';
    width: 12px;
    height: 6px;
    position: absolute;
    top: 2px;
    left: 1px;
    border: 2px solid #fff;
    border-top: none;
    border-right: none;
    transform: rotate(-45deg);
  }
}

.vnt-checkbox__text {
  padding: 0 9px;
  color: #000100;

  .vnt-checkbox__input:disabled ~ & {
    color: #999;
  }
}
</style>
