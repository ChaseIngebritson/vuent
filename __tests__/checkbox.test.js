import { mount } from '@vue/test-utils';
import Checkbox from '@/components/checkbox';

describe('Checkbox', () => {
  let wrapper;

  describe('by default', () => {

    beforeAll(() => {
      wrapper = mount(Checkbox);
    });

    test('is not disabled', () => {
      expect(wrapper.vm.disabled).toBe(false);
    });

    test('is not checked', () => {
      expect(wrapper.vm.checked).toBe(false);
    });

    test('label text is "Checkbox"', () => {
      expect(wrapper.vm.label).toBe('Checkbox');
    });

    test('renders correctly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

  });

  describe('can be disabled', () => {

    beforeAll(() => {
      wrapper = mount(Checkbox, {
        propsData: {
          disabled: true
        }
      });
    });

    test('renders correctly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    test('is disabled', () => {
      expect(wrapper.vm.disabled).toBe(true);
    });

  });

  describe('can be checked', () => {

    beforeAll(() => {
      wrapper = mount(Checkbox, {
        propsData: {
          checked: true
        }
      });
    });

    test('renders correctly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    test('is checked', () => {
      expect(wrapper.vm.checked).toBe(true);
    });

  });

  describe('can be checked and disabled', () => {

    beforeAll(() => {
      wrapper = mount(Checkbox, {
        propsData: {
          checked: true,
          disabled: true
        }
      });
    });

    test('renders correctly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    test('is checked', () => {
      expect(wrapper.vm.checked).toBe(true);
    });

    test('is disabled', () => {
      expect(wrapper.vm.disabled).toBe(true);
    });

  });

  test('can have content set using slot', () => {
    wrapper = mount(Checkbox, {
      slots: {
        default: '<span>Custom content</span>'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when not disabled, updates passed checked prop', () => {

    beforeAll(() => {
      wrapper = mount(Checkbox, {
        propsData: {
          checked: false
        }
      });

      const label = wrapper.find('label');
      label.trigger('click');
    });

    test('event is emitted', () => {
      expect(wrapper.emitted()['update:checked']).toBeDefined();
    });

    test('new checked state is emitted', () => {
      expect(wrapper.emitted()['update:checked'][0]).toEqual([true]);
    });

  });

  describe('when disabled, doesn\'t change state', () => {

    beforeAll(() => {
      wrapper = mount(Checkbox, {
        propsData: {
          checked: false,
          disabled: true
        }
      });

      const label = wrapper.find('label');
      label.trigger('click');
    });

    test('event is not emitted', () => {
      expect(wrapper.emitted()['update:checked']).toBeUndefined();
    });

  });

});
