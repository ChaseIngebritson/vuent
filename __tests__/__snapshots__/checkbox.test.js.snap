// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox by default renders correctly 1`] = `
<label class="vnt-checkbox">
  <input type="checkbox" class="vnt-checkbox__input"> <span class="vnt-checkbox__tick"></span> <span class="vnt-checkbox__text">Checkbox</span></label>
`;

exports[`Checkbox can be checked and disabled renders correctly 1`] = `
<label class="vnt-checkbox">
  <input type="checkbox" disabled="disabled" class="vnt-checkbox__input"> <span class="vnt-checkbox__tick"></span> <span class="vnt-checkbox__text">Checkbox</span></label>
`;

exports[`Checkbox can be checked renders correctly 1`] = `
<label class="vnt-checkbox">
  <input type="checkbox" class="vnt-checkbox__input"> <span class="vnt-checkbox__tick"></span> <span class="vnt-checkbox__text">Checkbox</span></label>
`;

exports[`Checkbox can be disabled renders correctly 1`] = `
<label class="vnt-checkbox">
  <input type="checkbox" disabled="disabled" class="vnt-checkbox__input"> <span class="vnt-checkbox__tick"></span> <span class="vnt-checkbox__text">Checkbox</span></label>
`;

exports[`Checkbox can have content set using slot 1`] = `
<label class="vnt-checkbox">
  <input type="checkbox" class="vnt-checkbox__input"> <span class="vnt-checkbox__tick"></span> <span class="vnt-checkbox__text"><span>Custom content</span></span>
</label>
`;
